src/include/alloc.h

```diff
@@ -106,22 +114,154 @@ niova_vbasic_nassigned(const struct niova_vbasic_allocator *nvba)
 }
 
 static inline int
-niova_vbasic_init(struct niova_vbasic_allocator *nvba,
-                  size_t region_size)
+niova_vbasic_init(struct niova_vbasic_allocator *nvba, size_t region_size)
 {
     if (!nvba)
         return -EINVAL;
 
     unsigned int unit_size = region_size / NIOVA_VBA_MAX_BITS;
-    if (unit_size == 0)
+
+    if (unit_size == 0 || unit_size > UINT32_MAX)
         return -EINVAL;
 
-    CONST_OVERRIDE(size_t, nvba->nvba_unit_size, unit_size);
+    CONST_OVERRIDE(uint32_t, nvba->nvba_unit_size, unit_size);
     nvba->nvba_bitmap = 0;
+    nvba->nvba_alignment = 0;
 
+    nvba->nvba_region_ptr = &nvba->nvba_region[0];
     return 0;
 }
 
+static inline int
+niova_vbasic_init_aligned(struct niova_vbasic_allocator *nvba,
+                          size_t region_size, size_t unit_size,
+                          uint32_t alignment)
+{
+    int rc = 0, err_loc = 0;
+    unsigned int nunits = 0;
+
+    if (!nvba)
+    {
+        err_loc = 1;
+        rc = -EINVAL;
+        goto xerror;
+    }
+
+    if (region_size == 0)
+    {
+        err_loc = 2;
+        rc = -ENODATA;
+        goto xerror;
+    }
+
+    if (region_size > NIOVA_VBA_REGSZ_MASK)
+    {
+        err_loc = 3;
+        rc = -EOVERFLOW;
+        goto xerror;
+    }
+
+    /* alignment is power of 2 */
+    if (alignment && number_of_ones_in_val32(alignment) != 1)
```

Comment:
I feel a bit turned upside down by this point.. I know we wanted to support `unit_size == 0` but how can `alignment == 0`?  If the user doesn't care about alignment then they must use `niova_vbasic_init()`.